### PR TITLE
[fix](cloud) Optimize prompt messages when no cluster

### PR DESCRIPTION
**Pull request by deardeng**
## Proposed changes

Issue Number: close #xxx

### CODE REVIEW

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Review by yujun777 — APPROVED** [review]
LGTM

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Comment by deardeng** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java [hunk 1]
```diff
@@ -90,7 +91,12 @@ public CloudRollupJobV2(String rawSql, long jobId, long dbId, long tableId, Stri
                 baseSchemaHash, rollupSchemaHash, rollupKeysType, rollupShortKeyColumnCount, origStmt);
         ConnectContext context = ConnectContext.get();
         if (context != null) {
-            String clusterName = context.getCloudCluster();
+            String clusterName = "";
+            try {
+                clusterName = context.getCloudCluster();
+            } catch (ComputeGroupException e) {
+                LOG.warn("failed to get cluster name", e);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java:98]: failed to get compute group name.

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java [hunk 2]
```diff
@@ -83,7 +84,12 @@ public CloudSchemaChangeJobV2(String rawSql, long jobId, long dbId, long tableId
         super(rawSql, jobId, dbId, tableId, tableName, timeoutMs);
         ConnectContext context = ConnectContext.get();
         if (context != null) {
-            String clusterName = context.getCloudCluster();
+            String clusterName = "";
+            try {
+                clusterName = context.getCloudCluster();
+            } catch (ComputeGroupException e) {
+                LOG.warn("failed to get cluster name", e);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java:91]: failed to get compute group name.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/CacheHotspotManagerUtils.java [hunk 3]
```diff
@@ -254,7 +255,11 @@ public AutoCloseConnectContext(ConnectContext connectContext) {
             this.previousContext = ConnectContext.get();
             this.connectContext = connectContext;
             connectContext.setThreadLocalInfo();
-            connectContext.getCloudCluster();
+            try {
+                connectContext.getCloudCluster();
+            } catch (ComputeGroupException e) {
+                LOG.warn("failed to get cloud cluster", e);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/CacheHotspotManagerUtils.java:261]: compute group name.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 4]
```diff
@@ -91,17 +93,26 @@ private boolean isColocated() {
         return Env.getCurrentColocateIndex().isColocateTable(tableId);
     }
 
-    private long getColocatedBeId(String cluster) {
+    private long getColocatedBeId(String cluster) throws ComputeGroupException {
         List<Backend> bes = ((CloudSystemInfoService) Env.getCurrentSystemInfo()).getBackendsByClusterId(cluster);
+        String clusterName = ((CloudSystemInfoService) Env.getCurrentSystemInfo()).getClusterNameByClusterId(cluster);
+        if (bes.isEmpty()) {
+            throw new ComputeGroupException(
+                String.format("There are no Backend nodes in the current cluster %s", clusterName),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:101]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 5]
```diff
@@ -91,17 +93,26 @@ private boolean isColocated() {
...
                 availableBes.add(be);
             }
         }
-        if (availableBes == null || availableBes.size() == 0) {
+
+        if (availableBes.isEmpty()) {
             LOG.warn("failed to get available be, clusterId: {}", cluster);
-            return -1;
+            throw new ComputeGroupException(
+                String.format("All the Backend nodes in the current cluster %s are in an abnormal state", clusterName),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:114]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 6]
```diff
@@ -129,7 +140,8 @@ public long getBackendId() {
                     ((CloudEnv) Env.getCurrentEnv()).checkCloudClusterPriv(cluster);
                 } catch (Exception e) {
                     LOG.warn("get cluster by session context exception");
-                    return -1;
+                    throw new ComputeGroupException(String.format("default cluster %s check auth failed", cluster),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:143]: default compute group.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 7]
```diff
@@ -139,35 +151,54 @@ public long getBackendId() {
...
                     LOG.debug("get cluster by context {}", cluster);
                 }
+                String clusterStatus = ((CloudSystemInfoService) Env.getCurrentSystemInfo())
+                        .getCloudStatusByName(cluster);
+                if (!Strings.isNullOrEmpty(clusterStatus)
+                        && Cloud.ClusterStatus.valueOf(clusterStatus)
+                        == Cloud.ClusterStatus.MANUAL_SHUTDOWN) {
+                    LOG.warn("auto start cluster {} in manual shutdown status", cluster);
+                    throw new ComputeGroupException(
+                        String.format("The current cluster %s has been manually shutdown", cluster),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:161]: current compute group.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 8]
```diff
@@ -139,35 +151,54 @@ public long getBackendId() {
...
         try {
-            cluster = ((CloudSystemInfoService) Env.getCurrentSystemInfo()).waitForAutoStart(cluster);
+            wakeUPCluster = ((CloudSystemInfoService) Env.getCurrentSystemInfo()).waitForAutoStart(cluster);
         } catch (DdlException e) {
             // this function cant throw exception. so just log it
             LOG.warn("cant resume cluster {}, exception", cluster, e);
         }
+        if (!Strings.isNullOrEmpty(wakeUPCluster) && !cluster.equals(wakeUPCluster)) {
+            cluster = wakeUPCluster;
+            LOG.warn("get backend input cluster {} useless, so auto start choose a new one cluster {}",
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:186]: new compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 9]
```diff
@@ -139,35 +151,54 @@ public long getBackendId() {
...
+                    cluster, wakeUPCluster);
+        }
         // check default cluster valid.
         if (Strings.isNullOrEmpty(cluster)) {
             LOG.warn("failed to get available be, clusterName: {}", cluster);
-            return -1;
+            throw new ComputeGroupException("cluster name is empty",
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:192]: compute group name is empty.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 10]
```diff
@@ -139,35 +151,54 @@ public long getBackendId() {
...
+                ComputeGroupException.FailedTypeEnum.CONNECT_CONTEXT_NOT_SET_COMPUTE_GROUP);
         }
         boolean exist = ((CloudSystemInfoService) Env.getCurrentSystemInfo())
                 .getCloudClusterNames().contains(cluster);
         if (!exist) {
             // can't use this default cluster, plz change another
             LOG.warn("cluster: {} is not existed", cluster);
-            return -1;
+            throw new ComputeGroupException(String.format("The current cluster %s is not registered in the system",
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:200]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 11]
```diff
@@ -257,10 +288,16 @@ private long getAvaliableBeId(String clusterId, Map<String, List<Long>> clusterT
...
 
-    public long hashReplicaToBe(String clusterId, boolean isBackGround) {
+    public long hashReplicaToBe(String clusterId, boolean isBackGround) throws ComputeGroupException {
         // TODO(luwei) list should be sorted
         List<Backend> clusterBes = ((CloudSystemInfoService) Env.getCurrentSystemInfo())
                 .getBackendsByClusterId(clusterId);
+        String clusterName = ((CloudSystemInfoService) Env.getCurrentSystemInfo()).getClusterNameByClusterId(clusterId);
+        if (clusterBes.isEmpty()) {
+            throw new ComputeGroupException(
+                String.format("There are no Backend nodes in the current cluster %s", clusterName),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:298]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 12]
```diff
@@ -272,11 +309,13 @@ public long hashReplicaToBe(String clusterId, boolean isBackGround) {
...
             }
         }
-        if (availableBes == null || availableBes.size() == 0) {
+        if (availableBes.isEmpty()) {
             if (!isBackGround) {
                 LOG.warn("failed to get available be, clusterId: {}", clusterId);
             }
-            return -1;
+            throw new ComputeGroupException(
+                String.format("All the Backend nodes in the current cluster %s are in an abnormal state", clusterName),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:317]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 13]
```diff
@@ -301,11 +340,17 @@ pickedBeId, getId(), partitionId, availableBes.size(), idx, index,
...
-    public List<Long> hashReplicaToBes(String clusterId, boolean isBackGround, int replicaNum) {
+    public List<Long> hashReplicaToBes(String clusterId, boolean isBackGround, int replicaNum)
+            throws ComputeGroupException {
         // TODO(luwei) list should be sorted
         List<Backend> clusterBes = ((CloudSystemInfoService) Env.getCurrentSystemInfo())
                 .getBackendsByClusterId(clusterId);
+        String clusterName = ((CloudSystemInfoService) Env.getCurrentSystemInfo()).getClusterNameByClusterId(clusterId);
+        if (clusterBes.isEmpty()) {
+            throw new ComputeGroupException(
+                String.format("There are no Backend nodes in the current cluster %s", clusterName),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:351]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java [hunk 14]
```diff
@@ -317,17 +362,19 @@ public List<Long> hashReplicaToBes(String clusterId, boolean isBackGround, int r
...
             }
         }
-        if (availableBes == null || availableBes.size() == 0) {
+        if (availableBes.isEmpty()) {
             if (!isBackGround) {
                 LOG.warn("failed to get available be, clusterId: {}", clusterId);
             }
-            return new ArrayList<Long>();
+            throw new ComputeGroupException(
+                String.format("All the Backend nodes in the current cluster %s are in an abnormal state", clusterName),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudReplica.java:370]: current compute group

File: fe/fe-core/src/main/java/org/apache/doris/cloud/load/CloudBrokerLoadJob.java [hunk 15]
```diff
@@ -88,7 +89,13 @@ public CloudBrokerLoadJob(long dbId, String label, BrokerDesc brokerDesc, Origin
     private void setCloudClusterId() throws MetaNotFoundException {
         ConnectContext context = ConnectContext.get();
         if (context != null) {
-            String clusterName = context.getCloudCluster();
+            String clusterName = "";
+            try {
+                clusterName = context.getCloudCluster();
+            } catch (ComputeGroupException e) {
+                LOG.warn("failed to get cluster name", e);
+                throw new MetaNotFoundException("failed to get cluster name " + e);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/load/CloudBrokerLoadJob.java:97]: compute group name

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java [hunk 16]
```diff
@@ -447,7 +453,14 @@ public HttpPut generateRequestForMySqlLoad(
     private String selectBackendForMySqlLoad(String database, String table) throws LoadException {
         Backend backend = null;
         if (Config.isCloudMode()) {
-            backend = StreamLoadHandler.selectBackend(ConnectContext.get().getCloudCluster());
+            String clusterName = "";
+            try {
+                clusterName = ConnectContext.get().getCloudCluster();
+            } catch (Exception e) {
+                LOG.warn("failed to get cloud cluster: " + e.getMessage());
+                throw new LoadException("failed to get cloud cluster: " + e);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/MysqlLoadManager.java:461]: compute group

File: fe/fe-core/src/main/java/org/apache/doris/qe/ConnectContext.java [hunk 17]
```diff
@@ -1272,6 +1249,11 @@ public String getCloudCluster(boolean updateErr) {
         if (!Strings.isNullOrEmpty(defaultCluster)) {
             cluster = defaultCluster;
             choseWay = "default cluster";
+            if (!Env.getCurrentEnv().getAuth().checkCloudPriv(getCurrentUserIdentity(),
+                    cluster, PrivPredicate.USAGE, ResourceTypeEnum.CLUSTER)) {
+                throw new ComputeGroupException(String.format("default cluster %s check auth failed", cluster),
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/qe/ConnectContext.java:1254]: compute group

File: fe/fe-core/src/main/java/org/apache/doris/qe/ConnectContext.java [hunk 18]
```diff
@@ -1282,10 +1264,13 @@ public String getCloudCluster(boolean updateErr) {
 
         if (Strings.isNullOrEmpty(cluster)) {
             LOG.warn("cant get a valid cluster for user {} to use", getCurrentUserIdentity());
+            ComputeGroupException exception = new ComputeGroupException(
+                    "the user is not granted permission to the cluster",
+                    ComputeGroupException.FailedTypeEnum.CURRENT_USER_NO_AUTH_TO_USE_ANY_COMPUTE_GROUP);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/qe/ConnectContext.java:1269]: compute group

File: fe/fe-core/src/main/java/org/apache/doris/qe/MasterOpExecutor.java [hunk 19]
```diff
@@ -211,7 +211,13 @@ private TMasterOpRequest buildStmtForwardParams() throws AnalysisException {
         params.setStmtId(ctx.getStmtId());
         params.setCurrentUserIdent(ctx.getCurrentUserIdentity().toThrift());
 
-        String cluster = ctx.getCloudCluster(false);
+        String cluster = "";
+        try {
+            cluster = ctx.getCloudCluster(false);
+        } catch (Exception e) {
+            LOG.warn("failed to get cloud cluster", e);
+            throw new AnalysisException("failed to get cloud cluster", e);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/qe/MasterOpExecutor.java:219]: compute group

File: regression-test/suites/cloud_p0/multi_cluster/test_no_cluster_hits.groovy [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+        sql """GRANT SELECT_PRIV ON *.*.* TO ${user}"""
+        try {
+            connectInDocker(user = user, password = '') {
+                // errCode = 2, detailMessage = the user is not granted permission to the cluster, ClusterException: CURRENT_USER_NO_AUTH_TO_USE_COMPUTE_GROUP
+                sql """select * from information_schema.columns"""
+            }
+        } catch (Exception e) {
+            logger.info("exception: {}", e.getMessage())
+            assertTrue(e.getMessage().contains("ClusterException: CURRENT_USER_NO_AUTH_TO_USE_ANY_COMPUTE_GROUP"))
+            assertTrue(e.getMessage().contains("the user is not granted permission to the cluster"))
```
Comment (dataroaring) [inline regression-test/suites/cloud_p0/multi_cluster/test_no_cluster_hits.groovy:66]: compute group

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Review by yujun777 — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**